ampd/src/lib.rs

```diff
@@ -40,7 +42,19 @@ const PREFIX: &str = "axelar";
 
 type HandlerStream<E> = Pin<Box<dyn Stream<Item = Result<Event, E>> + Send>>;
 
-pub async fn run(cfg: Config, state_path: PathBuf) -> Result<(), Error> {
+pub async fn run(cfg: Config, state: State) -> (Option<State>, Result<(), Error>) {
+    let app = prepare_app(cfg, state).await;
+
+    match app {
+        Ok(app) => app
+            .run()
+            .await
+            .then(|(state, report)| (Some(state), report)),
+        Err(err) => (None, Err(err)),
+    }
+}
```

Comment:
```suggestion
pub async fn run(cfg: Config, state: State) -> (State, Result<(), Error>) {
    let app = prepare_app(cfg, state.clone()).await;

    match app {
        Ok(app) => app.run().await.then(|(state, report)| (state, report)),
        Err(err) => (state, Err(err)),
    }
}
```
don't have a very strong opinion but this may also help simpliy code in main.rs. I'll leave it up to you.